client/components/phone-input/index.jsx

```diff
@@ -15,126 +17,125 @@ import { countries } from './data';
 
 const PhoneInput = React.createClass( {
 	propTypes: {
-		onChange: React.PropTypes.func,
-		initialValue: React.PropTypes.string,
-		selectedCountryCode: React.PropTypes.string.isRequired,
+		onChange: React.PropTypes.func.isRequired,
+		value: React.PropTypes.string.isRequired,
+		countryCode: React.PropTypes.string.isRequired,
 		countriesList: React.PropTypes.object.isRequired
 	},
 
-	getDefaultProps() {
-		return {
-			initialValue: '',
-			onChange: noop,
-		};
+	getCountry( countryCode = this.props.countryCode ) {
+		let selectedCountry = countries[ countryCode ];
+
+		if ( ! selectedCountry ) {
+			const data = find( this.props.countriesList.get() || [],
+				country => country.code.toLowerCase() === countryCode );
+			// Special cases where the country is in a disputed region and not globally recognized.
+			// At this point this should only be used for: Canary islands, Kosovo, Netherlands Antilles
+			if ( data ) {
+				selectedCountry = {
+					isoCode: countryCode,
+					dialCode: data.numeric_code.replace( '+', '' ),
+					nationalPrefix: ''
+				};
+			}
+		}
+		return selectedCountry;
 	},
 
 	getInitialState() {
-		const selectedCountry = countries[ this.props.selectedCountryCode ];
-		const formattedNumber = formatNumber( this.props.initialValue, selectedCountry );
 		return {
-			selectedCountry,
-			formattedNumber,
-			queryString: '',
 			freezeSelection: false
 		};
 	},
 
-	componentDidUpdate( prevProps, prevState ) {
-		if ( this.state.value && this.state.selectedCountry !== prevState.selectedCountry ) {
-			this.setState( {
-				formattedNumber: formatNumber( this.state.value, this.state.selectedCountry )
-			} );
+	componentWillReceiveProps( nextProps ) {
+		if ( this.props.value !== nextProps.value && nextProps.value ) {
+			const country = this.guessCountryFromValue( nextProps.value );
+			if ( country && country.isoCode !== this.getCountry().isoCode ) {
+				this.props.onChange( { countryCode: country.isoCode, value: nextProps.value } );
+			}
 		}
 	},
 
+	componentWillUpdate( nextProps ) {
+		const currentFormat = this.format( this.props ),
+			currentCursorPoint = this.numberInput.selectionStart,
+			nextFormat = this.format( nextProps );
+
+		let newCursorPoint = currentCursorPoint;
+		this.numberInput.value = nextFormat;
+
+		const nonDigitCountOld = currentFormat
+			.substring( 0, currentCursorPoint )
+			.split( '' )
+			.map( c => isNaN( parseInt( c, 10 ) ) )
```

Comment:
Could you use a more descriptive name here (and below)? I'm honestly not sure what `c` is supposed to represent here - it's coming from a `currentFormat` variable 😕 Cursor point? Character? :)